### ISSUE TITLE
Data Preprocessing Pipeline Implementation

### ISSUE THREAD

**Issue by Thamidu-Nadun** [Thamidu-Nadun/movie_recommendation_ml#3]
# Data Preprocessing Pipeline Implementation

## Overview
This issue outlines the complete data preprocessing pipeline required for the movie recommendation ML model. The preprocessing steps are crucial for data quality and model performance.

### 1. Remove Irrelevant Columns
- [x] Identify and analyze all columns in the dataset
- [x] Remove columns that don't contribute to movie recommendations
- [x] Document removed columns and reasoning
- [x] Examples of potentially irrelevant columns:
  - User registration timestamps (if not relevant)
  - Internal system IDs
  - Debug/logging columns
  - Redundant information

```python
# Example implementation
df_cleaned = df.drop(columns=['irrelevant_col1', 'irrelevant_col2'])
```

### 2. Handle Missing Values (Null Values)
- [x] Analyze missing value patterns across all columns
- [x] Implement appropriate strategies for each column type:
  - **Numerical columns**: Mean/median imputation or forward/backward fill
  - **Categorical columns**: Mode imputation or "Unknown" category
  - **Rating data**: Consider if missing means "not watched" vs actual missing data
- [x] Document imputation strategies used
- [x] Validate imputation results

```python
# Example strategies
df['age'].fillna(df['age'].median(), inplace=True)
df['genre'].fillna('Unknown', inplace=True)
df['rating'].fillna(df.groupby('user_id')['rating'].transform('mean'), inplace=True)
```

### 3. Convert Inconsistent Data (Date → Timestamp)
- [ ] Identify all date/time columns
- [ ] Standardize date formats to machine-readable timestamps
- [ ] Handle different date formats in the same column
- [ ] Extract useful features from dates (year, month, day of week, etc.)
- [ ] Convert to Unix timestamp or pandas datetime format

```python
# Example implementation
df['timestamp'] = pd.to_datetime(df['date_column']).astype('int64') // 10**9
df['year'] = pd.to_datetime(df['date_column']).dt.year
df['month'] = pd.to_datetime(df['date_column']).dt.month
```

### 4. Bucket into Groups (Age)
- [x] Analyze age distribution
- [x] Define meaningful age groups for movie preferences:
  - Children (0-12)
  - Teenagers (13-19)
  - Young Adults (20-35)
  - Middle-aged (36-50)
  - Seniors (51+)
- [x] Create age group categorical variable
- [x] Validate grouping makes sense for movie recommendations

```python
# Example implementation
def categorize_age(age):
    if age <= 12:
        return 'Children'
    elif age <= 19:
        return 'Teenagers'
    elif age <= 35:
        return 'Young_Adults'
    elif age <= 50:
        return 'Middle_Aged'
    else:
        return 'Seniors'

df['age_group'] = df['age'].apply(categorize_age)
```

### 5. Convert Categorical Data (Gender, Occupation)
- [x] **Gender encoding**:
  - Use label encoding or one-hot encoding
  - Handle non-binary/other categories appropriately
- [x] **Occupation encoding**:
  - Analyze occupation distribution
  - Consider grouping similar occupations
  - Use appropriate encoding method (label/ordinal/one-hot)
- [x] Document encoding schemes used
- [x] Save encoders for future use

```python
# Example implementations
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

# Label encoding for binary categories
le_gender = LabelEncoder()
df['gender_encoded'] = le_gender.fit_transform(df['gender'])

# One-hot encoding for multi-category
df_occupation = pd.get_dummies(df['occupation'], prefix='occupation')
df = pd.concat([df, df_occupation], axis=1)
```

### 6. Handle Outliers
- [x] Identify outliers in numerical columns using:
  - IQR method
  - Z-score method
  - Visualization (box plots, scatter plots)
- [x] Analyze outliers in context:
  - **Ratings**: Unusual rating patterns
  - **Age**: Unrealistic ages
  - **Movie data**: Extreme values
- [x] Choose appropriate outlier treatment:
  - Remove outliers
  - Cap/floor values
  - Transform data
- [ ] Document outlier handling decisions

```python
# Example outlier detection and handling
Q1 = df['rating'].quantile(0.25)
Q3 = df['rating'].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Remove outliers
df_no_outliers = df[(df['rating'] >= lower_bound) & (df['rating'] <= upper_bound)]

# Or cap outliers
df['rating_capped'] = df['rating'].clip(lower=lower_bound, upper=upper_bound)
```

### 7. Scaling
- [x] Identify numerical features that need scaling
- [ ] Choose appropriate scaling method:
  - **StandardScaler**: For normally distributed data
  - **MinMaxScaler**: For bounded ranges (0-1)
  - **RobustScaler**: When outliers are present
- [x] Apply scaling to training data
- [ ] Save scalers for test data transformation
- [x] Document scaling choices

```python
# Example scaling
from sklearn.preprocessing import StandardScaler, MinMaxScaler

scaler = StandardScaler()
numerical_cols = ['age', 'rating', 'timestamp']
df[numerical_cols] = scaler.fit_transform(df[numerical_cols])

# Save scaler for later use
import joblib
joblib.dump(scaler, 'scaler.pkl')
```

### 8. Feature Engineering (PCA → Correlation Matrix)
- [x] **Correlation Analysis**:
  - Calculate correlation matrix for all numerical features
  - Visualize correlations with heatmap
  - Identify highly correlated features (>0.8 or <-0.8)
  - Remove redundant features
- [x] **Principal Component Analysis (PCA)**:
  - Analyze explained variance ratio
  - Choose optimal number of components (e.g., 95% variance)
  - Apply PCA transformation
  - Document dimensionality reduction results
- [x] **Additional feature engineering**:
  - User-movie interaction features
  - Temporal features from timestamps
  - Genre preferences per user

```python
# Correlation analysis
import seaborn as sns
import matplotlib.pyplot as plt

correlation_matrix = df.corr()
plt.figure(figsize=(12, 8))
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', center=0)
plt.title('Feature Correlation Matrix')
plt.show()

# PCA implementation
from sklearn.decomposition import PCA

pca = PCA()
pca.fit(df_scaled)

# Plot explained variance
plt.plot(np.cumsum(pca.explained_variance_ratio_))
plt.xlabel('Number of Components')
plt.ylabel('Cumulative Explained Variance')
plt.title('PCA Analysis')
plt.show()

# Apply PCA with chosen components
n_components = 10  # Based on analysis
pca_final = PCA(n_components=n_components)
df_pca = pca_final.fit_transform(df_scaled)
```

## Acceptance Criteria
- [x] All preprocessing steps are implemented and tested
- [x] Code is well-documented with comments
- [x] Preprocessing pipeline is modular and reusable
- [ ] Data quality metrics are calculated before/after preprocessing
- [x] Visualizations are created to validate preprocessing steps
- [x] Preprocessed data is saved in appropriate format
- [x] All encoders/scalers are saved for production use

## Additional Notes
- Consider creating a preprocessing class/module for reusability
- Implement data validation checks after each step
- Create unit tests for preprocessing functions
- Document any assumptions made during preprocessing
- Consider the impact of preprocessing on model interpretability

## Expected Deliverables
1. Cleaned and preprocessed dataset
2. Preprocessing pipeline script/notebook
3. Documentation of all preprocessing decisions
4. Saved encoders, scalers, and transformers
5. Data quality report (before/after preprocessing)

---
**Priority**: High  
**Estimated Effort**: 2-3 days  
**Dependencies**: Raw dataset availability
